### ISSUE TITLE
Find a better solution for allowing nodeport ephemeral ports

### ISSUE THREAD

**Issue by cmische** [Misch-/Kubernetes-Capstone#4]
#3 Describes an issue where we had to open all kubernetes ephemeral ports to allow containers hosted on them to be accessible after an iptables DNAT created by kubernetes.

**Comment by cmische** [comment]
I got our firewall and security strategy nicely packaged up now.

After figuring out how to load balance with haproxy, the next step was to fix the fact that if our kubernetes iptables redirect to different ports, and you have your firewall on that host, you have to have both ports open or it will get blocked.

In our meeting last Wednesday, Matt mentioned that he had heard the firewall was supposed to be external, and that our issue is probably why. That made sense as to why nobody else was complaining about our problem of having to open all the ephemeral docker ports.

My solution was to sacrifice our third vm, and instead of deploying a kubernetes minion on it, I deployed haproxy and a firewall on it. That way I only have to open, lets say, port 80 on the proxy host, then it gets redirected and routed to kubernetes over an internal connection, and now that all external traffic for exposed services comes in through haproxy and then over the internal connection between nodes, I can just ifdown eth1 on both the kubernetes nodes so that they don't have access to the internet and turn off the firewall on both those nodes.

This means to ssh to kube1 and kube2 now, you must first ssh into the external ip of kube3, and then ssh into one of others through their internal address, as they are not directly accessible over the internet anymore, except through haproxy load balancing.

So now we have the hard external shell without the problems of having to place a firewall directly on the kubernetes nodes. If we every want to update packages or something, we can simply turn on the firewall, ifup eth1, or use a proxy or something.

Basically it's all fixed by having a seperate proxy host with a firewall on it for all external traffic.

```
root@kube03:~# ufw status verbose numbered
Status: active
Logging: on (low)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
22                         ALLOW IN    Anywhere
80                         ALLOW IN    Anywhere
22 (v6)                    ALLOW IN    Anywhere (v6)

root@kube03:~# 
```

Then we just have to open one port for each service we want to expose, otherwise, everything is only accessible internally.
